pkg/stream/pipeline.go

```diff
@@ -149,11 +151,17 @@ func (p *Pipeline) process(ctx context.Context, force bool) {
 	}()
 
 	p.ticker.Stop()
-	messages, err := p.callback.Process(ctx, p.batch)
 
-	if err != nil {
-		p.logger.Error(err, "could not process the batch")
-		return
+	var err error
+	var messages = p.batch
+
+	for _, stage := range p.stages {
+		messages, err = stage.Process(ctx, messages)
+
+		if err != nil {
+			p.logger.Error(err, "could not process the batch")
+			return
```

Comment:
as just discussed, the stage should remove invalid messages that should not be processed by the next stage. Due to that the return should also be removed